staging/src/k8s.io/client-go/transport/token_source.go

```diff
@@ -72,14 +92,20 @@ func NewCachedTokenSource(ts oauth2.TokenSource) oauth2.TokenSource {
 type tokenSourceTransport struct {
 	base http.RoundTripper
 	ort  http.RoundTripper
+	src  ResettableTokenSource
 }
 
 func (tst *tokenSourceTransport) RoundTrip(req *http.Request) (*http.Response, error) {
 	// This is to allow --token to override other bearer token providers.
 	if req.Header.Get("Authorization") != "" {
 		return tst.base.RoundTrip(req)
 	}
-	return tst.ort.RoundTrip(req)
+	resp, err := tst.ort.RoundTrip(req)
+	if err == nil && resp != nil && resp.StatusCode == 401 && tst.src != nil {
+		tst.src.ResetToken()
```

Comment:
The main thing I want to avoid here is races between concurrent requests hitting 401's, triggering refetching the token, then resetting a new valid token

Reply:
what do you think about adding singleflight or a timestamp for last successful request?

Reply:
timestamp seems ok. I would err on the side of not resetting a token and requiring more than one request to get a 401.

maybe something like:

```diff
diff --git a/staging/src/k8s.io/client-go/transport/token_source.go b/staging/src/k8s.io/client-go/transport/token_source.go
index d1e9b6e931d..ab3a11704a6 100644
--- a/staging/src/k8s.io/client-go/transport/token_source.go
+++ b/staging/src/k8s.io/client-go/transport/token_source.go
@@ -45,7 +45,7 @@ func TokenSourceWrapTransport(ts oauth2.TokenSource) func(http.RoundTripper) htt
 
 type ResettableTokenSource interface {
 	oauth2.TokenSource
-	ResetToken()
+	ResetTokenOlderThan(time.Time)
 }
 
 // ResettableTokenSourceWrapTransport returns a WrapTransport that injects bearer tokens
@@ -100,9 +100,10 @@ func (tst *tokenSourceTransport) RoundTrip(req *http.Request) (*http.Response, e
 	if req.Header.Get("Authorization") != "" {
 		return tst.base.RoundTrip(req)
 	}
+	start := time.Now()
 	resp, err := tst.ort.RoundTrip(req)
 	if err == nil && resp != nil && resp.StatusCode == 401 && tst.src != nil {
-		tst.src.ResetToken()
+		tst.src.ResetTokenOlderThan(start)
 		return tst.ort.RoundTrip(req)
 	}
 	return resp, err
@@ -145,6 +146,7 @@ type cachingTokenSource struct {
 
 	sync.RWMutex
 	tok *oauth2.Token
+	t   time.Time
 
 	// for testing
 	now func() time.Time
@@ -177,12 +179,16 @@ func (ts *cachingTokenSource) Token() (*oauth2.Token, error) {
 		return ts.tok, nil
 	}
 
+	ts.t = ts.now()
 	ts.tok = tok
 	return tok, nil
 }
 
-func (ts *cachingTokenSource) ResetToken() {
+func (ts *cachingTokenSource) ResetTokenOlderThan(t time.Time) {
 	ts.Lock()
 	defer ts.Unlock()
-	ts.tok = nil
+	if ts.t.Before(t) {
+		ts.tok = nil
+		ts.t = time.Time{}
+	}
 }
```

Reply:
did you mean
```golang
 if err == nil && resp != nil && resp.StatusCode == 401 && tst.src != nil && tst.src.ResetTokenOlderThan(start) {
    return tst.ort.RoundTrip(req)
 }
```
by `I would err on the side of not resetting a token and requiring more than one request to get a 401.`?

if we decide not to retry request, we'd just return no matter reset or not.

Reply:
no, I meant that by capturing `start` prior to making the request, it is possible a just-acquired-and-immediately-invalid token would not be reset by `ResetTokenOlderThan(start)`, and it would have to wait for a subsequent RoundTrip request to fail before it would be reset

Reply:
makes sense. now in unauthorized case, `Token` would only be called once. see unit tests.